libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,312 @@
+    static_assert(is_same_v<decay_t<_Func>, _Func>);
+    __construct<_Func>(std::forward<_Args>(__args)...);
+  }
+
+  template <class _Func, class _InitListType, class... _Args>
+    requires is_constructible_v<decay_t<_Func>, initializer_list<_InitListType>, _Args...> && __is_callable_from<_Func>
+  _LIBCPP_HIDE_FROM_ABI explicit move_only_function(
+      in_place_type_t<_Func>, initializer_list<_InitListType> __il, _Args&&... __args) {
+    __construct<_Func>(__il, std::forward<_Args>(__args)...);
+  }
+
+  // TODO: Do we want to make this `noexcept` as an extensions?
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(move_only_function&& __other) noexcept {
+    move_only_function(std::move(__other)).swap(*this);
+    return *this;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(nullptr_t) noexcept {
+    __reset();
+    return *this;
+  }
+
+  // TODO: Do we want to make this `noexcept` as an extensions?
+  template <class _Func>
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(_Func&& __func) noexcept {
+    move_only_function(std::forward<_Func>(__func)).swap(*this);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI ~move_only_function() { __reset(); }
+
+  // [func.wrap.move.inv]
+  _LIBCPP_HIDE_FROM_ABI explicit operator bool() const noexcept { return __call_ != nullptr; }
+
+  _LIBCPP_HIDE_FROM_ABI _ReturnT operator()(_ArgTypes... __args) _LIBCPP_MOVE_ONLY_FUNCTION_CVREF
+      noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) {
+    __call_(__buffer_.data(), std::forward<_ArgTypes>(__args)...);
+  }
+
+  // [func.wrap.move.util]
+  _LIBCPP_HIDE_FROM_ABI void swap(move_only_function& __other) noexcept {
+    std::swap(__call_, __other.__call_);
+    std::swap(__destroy_, __other.__destroy_);
+    std::swap(__buffer_, __other.__buffer_);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend void swap(move_only_function& __lhs, move_only_function& __rhs) noexcept {
+    __lhs.swap(__rhs);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI friend bool operator==(const move_only_function& __func, nullptr_t) noexcept { return !__func; }
+
+private:
+  using _CallFn    = _ReturnT(_LIBCPP_MOVE_ONLY_FUNCTION_CV std::byte*, _ArgTypes...);
+  using _DestroyFn = void(void*);
+
+  static constexpr size_t __buffer_size_ = 4 * sizeof(void*);
+
+  _CallFn* __call_ = nullptr;
+  _DestroyFn* __destroy_ = nullptr;
+  alignas(void*) array<byte, __buffer_size_> __buffer_;
+
+  template <class _Func>
+  static constexpr bool __fits_in_buffer_impl =
+      __libcpp_is_trivially_relocatable_v<_Func> && sizeof(_Func) <= __buffer_size_ &&
+      alignof(_Func) <= alignof(void*);
+
+  template <class _Func>
+  static constexpr bool __fits_in_buffer = __fits_in_buffer_impl<remove_cvref_t<_Func>>;
+};
+
+template <class _ReturnT, class... _Args>
+struct __libcpp_is_trivially_relocatable<move_only_function<_ReturnT(
+    _Args...) _LIBCPP_MOVE_ONLY_FUNCTION_CVREF noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT)>> : true_type {};
```

Comment:
It has only now occurred to me that this partial specialization of `__libcpp_is_trivially_relocatable` is useless and can be removed (hooray!). You only ever test `__libcpp_is_trivially_relocatable<T>` when you're deciding whether to store a `T` inside a `move_only_function`, and obviously you will never store a `move_only_function` inside another `move_only_function` (because it'll be too big to fit in the buffer). Therefore, this partial specialization isn't serving any purpose, and so we needn't feel any qualms about eliminating it. :)

I can't off the top of my head think of any libc++-defined types that are callable, and trivially relocatable, and _not_ trivially copyable, except for the `std::function`-esque ones that are too big to fit in the buffer anyway. (`reference_wrapper<int()>` is already trivially copyable.) There are tons of such types in P1144 — e.g. `decltype([s = std::string("")]() { return s; })` — but you physically can't write a partial specialization for those types; you need P1144 in order to get that performance.